_posts/2024-07-03-oidc-cient-request.adoc

```diff
@@ -0,0 +1,440 @@
+
+* *Public Key:* This is the part of the keypair that is shared with trusted parties. Typically, the OpenID provider shared a set of public keys with us which we use to encrypt our JWT's among other things. Public keys can be extracted from certificates.
+
+* *Certificate:* This is used to verify the signatures among other things. We share the certificates of keypairs we generate with the OpenID provider, and it contains the public key.
+
+Keypairs with public and private keys are known as *asymmetric* keys. Some keys are *symmetric*, meaning they only have 1 secret key that is used by both the client and the server. Each key has an algorithm that is used to generate it, such as, RSA, Elliptic Curve, etc. An example of a symmetric algorithm is HS256, where the same key is used to sign a JWT and verify the signature.
+
+=== Generating Key Pairs
+
+We can use the `keytool` tool in the cli to generate our keys and store them in a keystore file. We can generate keystore files of type `JKS` and `PKCS12` for our application. The general format for doing this is as follows:
+```
+    $ keytool -genkeypair -alias <keystore_alias> -keyalg <algorithm> -keysize <key_size> -validity <validity_in_days>
+-keystore <keystore_name> -dname "<distinguished_name>" -keypass <private_key_password> -storepass <keystore_password>
+```
+
+==== Generating PKCS12 Keystore
+
+Let us first create a PKCS12 keystore file with 2 different keys. Let's first generate an RSA keyPair: 
+```
+    $ keytool -genkeypair -alias rsaKey -keyalg RSA -keysize 2048 -validity 365 -keystore Keycloak.keystore.pkcs12 -dname "CN=client" -keypass password -storepass password
+```
+And for Elliptical Curve keys, use the following commands: 
+```
+    $ keytool -genkeypair -alias ecKey -keyalg EC -groupname secp256r1 -validity 365 -keystore Keycloak.keystore.pkcs12 -dname "CN=client" -keypass password -storepass password
+``` 
+Note that in this case, we have specified the *key size* for the RSA keys, and the *groupname* for Elliptical Curve keys.
+For the RSA keys, the same keysize can be used for RSA algorithms with different SHA values (i.e. RSA256, RSA512).
+However, this is not the case for Elliptical Curve keys. For ES256 algorithms, you need to use the groupname associated with the SHA value (i.e. `-groupname secp256r1` and for ES384, use `-groupname secp384r1` and so on). You can use the following commands to see the contents of the keystore: 
+```
+    $ keytool -list -v -keystore Keycloak.keystore.pkcs12 -storepass password
+```
+
+If you would like to use JKS keystore files instead, see https://docs.redhat.com/en/documentation/red_hat_jboss_data_virtualization/6.4/html/security_guide/create_a_privatepublic_key_pair_with_keytool[here] for more information.
+
+== Log Into the OpenShift Cluster
+Before we can deploy our application, we need to log into an OpenShift cluster. You can log in via the OpenShift CLI:
+```
+oc login -u myUserName
```

Comment:
Upon login a list of available projects are listed.   The user must designate the project to make active with cmd,
   oc project <PROJECT_NAME>

Reply:
Added a note for this